README.md
# NVIspatial: Tools to facilitate working with Spatial Data at NVI

<!-- README.md is generated from README.Rmd. Please edit that file -->

-   [Overview](#overview)
-   [Installation](#installation)
-   [Usage](#usage)
-   [Copyright and license](#copyright-and-license)
-   [Contributing](#contributing)

## Overview

`NVIspatial`provides tools to facilitate working with spatial data at
NVI. Currently, NVI spatial holds the function ‘is\_it\_in\_Norway’. The
package will be extended with other tools for handling (Norwegian)
spatial data.

`NVIspatial` is part of `NVIverse`, a collection of R-packages with
tools to facilitate data management and data reporting at the Norwegian
Veterinary Institute (NVI). The `NVIverse` consists of the following
packages: `NVIconfig`, `NVIdb`, `NVIspatial`, `NVIpretty`, `NVIbatch`,
`OKplan`, `OKcheck`, `NVIcheckmate`, `NVIpackager`, `NVIrpackages`. See
[Contribute to
NVIspatial](https://github.com/NorwegianVeterinaryInstitute/NVIspatial/blob/main/CONTRIBUTING.md)
for more information.

## Installation

`NVIspatial` is available at
[GitHub](https://github.com/NorwegianVeterinaryInstitute). To install
`NVIspatial` you will need:

-   R version > 4.0.0
-   R package `remotes`
-   Rtools 4.0 or Rtools 4.2 depending on R version

First install and attach the `remotes` package.

    install.packages("remotes")
    library(remotes)

To install (or update) the `NVIspatial` package, run the following code:

    remotes::install_github("NorwegianVeterinaryInstitute/NVIspatial",
        upgrade = FALSE,
        build = TRUE,
        build_vignettes = TRUE)

## Usage

The `NVIspatial` package needs to be attached.

    library(NVIspatial)

`NVIspatial` provides tools to facilitate working with spatial data at
NVI. Currently, NVI spatial holds the function ‘is\_it\_in\_Norway’. The
package will be extended with other tools for handling (Norwegian)
spatial data.

The full list of all available functions and datasets can be accessed by
typing

    help(package = "NVIspatial")

Please check the NEWS for information on new features, bug fixes and
other changes.

## Copyright and license

Copyright (c) 2022 Norwegian Veterinary Institute.  
Licensed under the BSD\_3\_clause License. See
[License](https://github.com/NorwegianVeterinaryInstitute/NVIspatial/blob/main/LICENSE)
for details.

## Contributing

Contributions to develop `NVIspatial` is highly appreciated. There are
several ways you can contribute to this project: ask a question, propose
an idea, report a bug, improve the documentation, or contribute code.
See [Contribute to
NVIspatial](https://github.com/NorwegianVeterinaryInstitute/NVIspatial/blob/main/CONTRIBUTING.md)
for more information.

------------------------------------------------------------------------

<!-- Code of conduct -->

Please note that the NVIspatial project is released with a [Contributor
Code of
Conduct](https://github.com/NorwegianVeterinaryInstitute/NVIspatial/blob/main/CODE_OF_CONDUCT.md).
By contributing to this project, you agree to abide by its terms.


NEWS
NVIspatial 0.1.0.9000 - (2023-##-##)
----------------------------------------

New features:

-


Bug fixes:

-


Other changes:

-


BREAKING CHANGES:

-


NVIspatial 0.1.0 - (2022-08-21)
-------------------------------

First release:

  Tools to facilitate working with Spatial Data at NVI

- is_it_in_Norway checks if geo locations are within Norway and set non-Norwegian coordinates to missing and/or flag them.


tests/testthat.R
library(testthat)
library(NVIspatial)

test_check("NVIspatial")


notes/test_Prodtilskudd.R
#### SET UP OF R-ENVIRONMENT ----
# Packages
library(NVIbatch)
# use_pkg("plyr")
use_pkg(pkg = c("dplyr", "geosphere", "openxlsx", "sf", "tmap"))
use_NVIverse(pkg = c("NVIdb", "NVIspatial"))

# GLOBAL VARIABLES
# Todays date in the  format yyyymmdd to be used in filenames
today<-format(Sys.Date(),"%Y%m%d")

# Load data.frame exported from PJS if not already in memory
load(file = paste0(set_dir_NVI("Prodregister"),"Radata/PJSprodusenter.Rdata") )
prodnr_2_current_prodnr <- read_prodnr_2_current_prodnr()
PTdata <- read_Prodtilskudd()
PTprodusent <- PTdata[, c("gjeldende_prodnr8", "Orgnavn")]
PTprodusent <- unique(PTprodusent)

#### SELECT ONE SET OF COORDINATES FOR Prodnr8 WHEN < 1 m DIFFERENCE IN DISTANCE BETWEEN POINTS ----
# Generates table with unique combinations of Prodnr8 and coordinates, only
Koordinater <- PJSprodusenter %>%
  select(identifikator, geo_eu89_o, geo_eu89_n) %>%
  mutate(identifikator = substr(identifikator, 1, 8)) %>%
  arrange(identifikator) %>%
  distinct()

Koordinater <- is_it_in_Norway(data = Koordinater,
                               shape = Norway_with_buffer_100,
                               coordinates = c("geo_eu89_o", "geo_eu89_n"),
                               valid = "valid_buffer_100",
                               set_to_NA = FALSE)

Koordinater <- is_it_in_Norway(data = Koordinater,
                               shape = Norway_with_buffer_500,
                               coordinates = c("geo_eu89_o", "geo_eu89_n"),
                               valid = "valid_buffer_500",
                               set_to_NA = FALSE)

ktr1 <- subset(Koordinater, Koordinater$valid_buffer_500 == 0)
ktr2 <- subset(ktr1, !is.na(ktr1$geo_eu89_o) & ktr1$geo_eu89_o >1 &
                 !is.na(ktr1$geo_eu89_n) & ktr1$geo_eu89_n >1)

ktr3 <- ktr2 %>%
  add_produsent_properties(translation_table = prodnr_2_current_prodnr,
                           code_column = c("identifikator" = "prodnr8"),
                           new_column = "gjeldende_prodnr8",
                           impute_old_when_missing = TRUE) %>%
  left_join(PTprodusent, by = "gjeldende_prodnr8") %>%
  filter(!is.na(Orgnavn)) %>%
    st_as_sf(coords = c("geo_eu89_o", "geo_eu89_n"), crs = 4326)

tmap_mode("view")
tm_shape(Norway_with_buffer_500) +
  tm_borders() +
  tm_shape(ktr3) +
  tm_dots(col = "blue")




tests/testthat/test_is_it_in_Norway.R
library(NVIspatial)
library(testthat)
library(tibble)

test_that("is_it_in_Norway Mark geolocations outside Norway", {
  geolocations <- data.frame(
    "eier_lokalitetnr" = c("1", "2", "3"),
    "geo_eu89_o" = c("5.3105549", "0.3105549", NA),
    "geo_eu89_n" = c("60.3551767", "60.3551767", NA))

  correct_result <- cbind(geolocations,
                          "valid_coordinate" = c(1, 0, 0))

  expect_identical(is_it_in_Norway(data = geolocations,
                                   coordinates = c("geo_eu89_o", "geo_eu89_n"),
                                   polygon = Norway_with_buffer_500,
                                   valid = TRUE,
                                   set_to_NA = FALSE),
                   correct_result)


  correct_result <- cbind(geolocations,
                          "valid_coordinate_500" = c(1, 0, 0))

  geolocations <- is_it_in_Norway(data = geolocations,
                                  coordinates = c("geo_eu89_o", "geo_eu89_n"),
                                  polygon = Norway_with_buffer_500,
                                  valid = "valid_coordinate_500",
                                  set_to_NA = FALSE)

  expect_identical(geolocations, correct_result)

  expect_identical(is_it_in_Norway(data = geolocations,
                                   coordinates = c("geo_eu89_o", "geo_eu89_n"),
                                   polygon = Norway_with_buffer_500,
                                   valid = "valid_coordinate_500",
                                   set_to_NA = FALSE),
                   correct_result)
})

test_that("is_it_in_Norway Set to NA for geolocations outside Norway", {
  geolocations <- data.frame("eier_lokalitetnr" = c("1", "2", "3"),
                             "geo_eu89_o" = c("5.3105549", "0.3105549", NA),
                             "geo_eu89_n" = c("60.3551767", "60.3551767", NA))

  correct_result <- data.frame("eier_lokalitetnr" = c("1", "2", "3"),
                               "geo_eu89_o" = c("5.3105549", NA, NA),
                               "geo_eu89_n" = c("60.3551767", NA, NA))

  expect_identical(is_it_in_Norway(data = geolocations,
                                   coordinates = c("geo_eu89_o", "geo_eu89_n"),
                                   polygon = Norway_with_buffer_500,
                                   valid = FALSE,
                                   set_to_NA = TRUE),
                   correct_result)
})


test_that("is_it_in_Norway data as tibble", {
  geolocations <- tibble("eier_lokalitetnr" = c("1", "2", "3"),
                         "geo_eu89_o" = c("5.3105549", "0.3105549", NA),
                         "geo_eu89_n" = c("60.3551767", "60.3551767", NA))

  correct_result <- tibble("eier_lokalitetnr" = c("1", "2", "3"),
                           "geo_eu89_o" = c("5.3105549", NA, NA),
                           "geo_eu89_n" = c("60.3551767", NA, NA))

  expect_identical(is_it_in_Norway(data = geolocations,
                                   coordinates = c("geo_eu89_o", "geo_eu89_n"),
                                   polygon = Norway_with_buffer_500,
                                   valid = FALSE,
                                   set_to_NA = TRUE),
                   correct_result)
})


test_that("is_it_in_Norway error testing", {
  linewidth <- options("width")
  options(width = 80)

  geolocations <- c("eier_lokalitetnr" = c("1", "2", "3"),
                    "geo_eu89_o" = c("5.3105549", "0.3105549", NA),
                    "geo_eu89_n" = c("60.3551767", "60.3551767", NA))


  expect_error(is_it_in_Norway(data = geolocations,
                               coordinates = c("geo_eu89_o", "geo_eu89_n"),
                               polygon = Norway_with_buffer_500,
                               valid = FALSE,
                               set_to_NA = TRUE),
               regexp = "Must be of type 'data.frame'")

  geolocations <- data.frame("eier_lokalitetnr" = c("1", "2", "3"),
                             "geo_eu89_o" = c("5.3105549", "0.3105549", NA),
                             "geo_eu89_n" = c("60.3551767", "60.3551767", NA))
  expect_error(is_it_in_Norway(data = geolocations,
                               coordinates = c("x", "y"),
                               polygon = Norway_with_buffer_500,
                               valid = FALSE,
                               set_to_NA = TRUE),
               regexp = "'coordinates': Must be a subset of", fixed = TRUE)

  expect_error(is_it_in_Norway(data = geolocations,
                               coordinates = c("geo_eu89_o", "geo_eu89_n"),
                               polygon = "Norway_with_buffer_500",
                               valid = FALSE,
                               set_to_NA = TRUE),
               regexp = "Must be of type 'data.frame'")

  expect_error(is_it_in_Norway(data = geolocations,
                               coordinates = c("geo_eu89_o", "geo_eu89_n"),
                               polygon = Norway_with_buffer_500,
                               valid = geolocations,
                               set_to_NA = TRUE),
               regexp = "'valid': One of the following must apply:", fixed = TRUE)

  expect_error(is_it_in_Norway(data = geolocations,
                               coordinates = c("geo_eu89_o", "geo_eu89_n"),
                               polygon = Norway_with_buffer_500,
                               valid = FALSE,
                               set_to_NA = "TRUE"),
               regexp = "'set_to_NA': Must be of type 'logical flag'", fixed = TRUE)

  options(width = unlist(linewidth))
})


R/data.R
#' @title Geographical outline of Norway with 500 m buffer.
#'
#' @description A special features object describing a polygon with the geographical
#'     outline of Norway mainland with a buffer of 500 meter.
#' @details The polygon is used as input to `is_it_in_Norway` to check if a geo
#'     location is situated on land in Norway. The outline of Norway mainland has
#'     been increased with a buffer of 500 m to ensure that that minor inaccuracies
#'     in the geo locations do not invalidate the geo coordinates. A test using
#'     farmer locations in Norway shows that for this purpose, a buffer of 500 m
#'     is suited to differentiate between farmers with geo coordinates on
#'     Norwegian mainland or outside.
#'
#' The outline of Norway is constructed by combining  the country border and
#'     the coast line into a polygon. The resolution of the source files is based
#'     on the N500 series.
#'
#' @source "./data-raw/generate_Norway_with_buffer" in package `NVIspatial`.
"Norway_with_buffer_500"


data-raw/generate_Norway_with buffer.R
library(magrittr)
library(NVIdb)

filename <- "" # fill in pathname
Norway <- sf::st_read(filename)

proj_str <- 'PROJCS["ETRS89 / UTM zone 33N",GEOGCS["ETRS89",
                            DATUM["European_Terrestrial_Reference_System_1989",
                            SPHEROID["GRS 1980",6378137,298.257222101,
                            AUTHORITY["EPSG","7019"]],
                            TOWGS84[0,0,0,0,0,0,0],
                            AUTHORITY["EPSG","6258"]],
                            PRIMEM["Greenwich",0,
                            AUTHORITY["EPSG","8901"]],
                            UNIT["degree",0.0174532925199433,
                            AUTHORITY["EPSG","9122"]],
                            AUTHORITY["EPSG","4258"]],
                            PROJECTION["Transverse_Mercator"],
                            PARAMETER["latitude_of_origin",0],
                            PARAMETER["central_meridian",15],
                            PARAMETER["scale_factor",0.9996],
                            PARAMETER["false_easting",500000],
                            PARAMETER["false_northing",0],
                            UNIT["metre",1,
                            AUTHORITY["EPSG","9001"]],
                            AXIS["Easting",EAST],
                            AXIS["Northing",NORTH],
                            AUTHORITY["EPSG","25833"]]'

Norway_with_buffer_500 = Norway %>%
  sf::st_transform(crs = proj_str) %>%
  sf::st_buffer(500)

# SAVE IN PACKAGE DATA ----
usethis::use_data(name = Norway_with_buffer_500, overwrite = TRUE, internal = FALSE, compress = "xz")


R/is_it_in_Norway.R
#' @title Checks if the geo-location is within Norway
#' @description Checks if the geo location of a pair of geo coordinates, is within Norway.
#'
#' @details AS default, the check is performed towards a polygon with the shape
#'     of the Norwegian land area with buffer of 500 m. The polygon is based on
#'     the Norwegian country borders and coastline in N500. This should be sufficient
#'     accurate for checking farms, but will not be sufficient for checking all
#'     kinds of buildings, in particular buildings that can be expected to be
#'     situated on small islands.
#'
#' The check is intended for geo-locations on land, not in the sea. However,
#'     by changing the input to polygon, the check can be performed towards other
#'     areas. The input needs to be a special features object.
#'
#' To return a data frame with the invalid geo coordinates
#'     set to NA, use `valid = FALSE` and `set_to_NA = TRUE`. If you
#'     would like to inspect the invalid geo coordinates, use `valid = TRUE`
#'     and `set_to_NA = FALSE`.
#'
#' @param data \[\code{data.frame}\]\cr
#' Input data with location information that should be checked.
#' @param coordinates \[\code{character}\]\cr
#' Vector with the column names or the position numbers of the columns with the
#'     geo coordinates in data. The order of the geo coordinates should be given
#'     as c("longitude", "latitude").
#' @param input_projection the projection of the geo locations that should be checked.
#' @param polygon \[\code{sf}\]\cr
#' The polygon that you want to check if the geo locations are
#'     within. The polygon must be given as a special feature object. Defaults to
#'     `Norway_with_buffer_500`.
#' @param polygon_projection Projection of the data defining the polygon. Defaults
#'     to the projection of `Norway_with_buffer_500`.
#' @param valid \[\code{logical} or \code{character}\]\cr
#' Should the results include a variable with information on whether the
#'     geo location were within the polygon? Defaults to `FALSE`. The name of the
#'     new column defaults to "valid_coordinate". If another name is wanted, you
#'     can use the preferred column name as input argument.
#' @param set_to_NA \[\code{logical}\]\cr
#' Should the invalid geo coordinates be set to `NA`? Defaults to `TRUE`.
#'
#' @return A data frame. If `set_to_NA = TRUE` the geo coordinates that is
#'     outside Norway is set to `NA`. If `valid = TRUE`, a column named
#'     "valid_coordinate" (or the name you choose), is added and the observations
#'     are given the value 1 if the coordinate is within Norway and 0 if not.
#'
#' @export
#' @examples
#' \dontrun{
#' library(NVIspatial)
#' library(dplyr)
#' library(sf)
#'
#' geolocations <- data.frame(
#'   "eier_lokalitetnr" = c("1", "2", "3"),
#'   "geo_eu89_o" = c("5.3105549", "0.3105549", NA),
#'   "geo_eu89_n" = c("60.3551767", "60.3551767", NA))
#'
#' geolocations <- is_it_in_Norway(data = geolocations,
#'                                 coordinates = c("geo_eu89_o", "geo_eu89_n"),
#'                                 polygon = Norway_with_buffer_500,
#'                                 valid = TRUE,
#'                                 set_to_NA = FALSE)
#'
#' geolocations <- is_it_in_Norway(data = geolocations,
#'                                 coordinates = c("geo_eu89_o", "geo_eu89_n"),
#'                                 polygon = Norway_with_buffer_500,
#'                                 valid = "valid_coordinate_500",
#'                                 set_to_NA = FALSE)
#' }
#'
is_it_in_Norway <- function(data,
                            coordinates,
                            input_projection = 4326,
                            polygon = NVIspatial::Norway_with_buffer_500,
                            polygon_projection = paste('PROJCS["ETRS89 / UTM zone 33N",GEOGCS["ETRS89",',
                                                       'DATUM["European_Terrestrial_Reference_System_1989",',
                                                       'SPHEROID["GRS 1980",6378137,298.257222101,',
                                                       'AUTHORITY["EPSG","7019"]],',
                                                       'TOWGS84[0,0,0,0,0,0,0],',
                                                       'AUTHORITY["EPSG","6258"]],',
                                                       'PRIMEM["Greenwich",0,',
                                                       'AUTHORITY["EPSG","8901"]],',
                                                       'UNIT["degree",0.0174532925199433,',
                                                       'AUTHORITY["EPSG","9122"]],',
                                                       'AUTHORITY["EPSG","4258"]],',
                                                       'PROJECTION["Transverse_Mercator"],',
                                                       'PARAMETER["latitude_of_origin",0],',
                                                       'PARAMETER["central_meridian",15],',
                                                       'PARAMETER["scale_factor",0.9996],',
                                                       'PARAMETER["false_easting",500000],',
                                                       'PARAMETER["false_northing",0],',
                                                       'UNIT["metre",1,',
                                                       'AUTHORITY["EPSG","9001"]],',
                                                       'AXIS["Easting",EAST],',
                                                       'AXIS["Northing",NORTH],',
                                                       'AUTHORITY["EPSG","25833"]]'),
                            valid = FALSE,
                            set_to_NA = TRUE) {

  # ARGUMENT CHECKING ----
  # Object to store check-results
  checks <- checkmate::makeAssertCollection()
  # Perform checks

  # data
  checkmate::assert_data_frame(data, null.ok = FALSE, add = checks)
  # coordinates
  checkmate::assert_subset(coordinates, choices = colnames(data), empty.ok = FALSE, add = checks)
  # input_projection
  checkmate::assert(checkmate::check_integerish(input_projection, lower = 0, upper = 999999, len = 1, any.missing = FALSE),
                    checkmate::check_string(input_projection),
                    combine = "or",
                    add = checks)
  # shape
  checkmate::assert_data_frame(polygon, null.ok = FALSE, add = checks)
  # shape_projection
  checkmate::assert(checkmate::check_integerish(polygon_projection, lower = 0, upper = 999999, len = 1, any.missing = FALSE),
                    checkmate::check_string(polygon_projection),
                    combine = "or",
                    add = checks)
  # valid
  checkmate::assert(checkmate::check_flag(valid),
                    checkmate::check_character(valid, min.chars = 1, len = 1, any.missing = FALSE),
                    combine = "or",
                    add = checks)
  # set_to_NA
  checkmate::assert_flag(set_to_NA, add = checks)

  # Report check-results
  checkmate::reportAssertions(checks)

  valid_name <- "valid_coordinate"
  if (class(valid) == "character") {
    valid_name <- valid
    valid <- TRUE
  }
  # FIND GEO-LOCATIONS WITHIN NORWAY ----
  # Make new data frame with geo-coordinates to generate spatial information, transform missing to 0
  xydata <- as.data.frame(data[, coordinates])
  xydata[which(is.na(xydata[, 1])), 1] <- 0
  xydata[which(is.na(xydata[, 2])), 2] <- 0

  # Transform points to geographical information
  xydata <- sf::st_as_sf(xydata, coords = c(1, 2), crs = sf::st_crs(input_projection))
  xydata <- sf::st_transform(xydata, crs = polygon_projection)

  # Check if points within polygon
  xydata$valid_coordinate <- sf::st_intersects(xydata, polygon)
  xydata$valid_coordinate <- lengths(xydata$valid_coordinate)
  xydata[which(xydata$valid_coordinate > 0), "valid_coordinate"] <- 1
  ## in case the polygon file should include geo-location (0, 0)
  xydata[which(is.na(data[, coordinates[1]]) | is.na(data[, coordinates[2]])), "valid_coordinate"] <- 0

  # REPORTS RESULT IN ACCORD WITH INPUT ----
  if (set_to_NA == TRUE) {
    data[which(xydata$valid_coordinate == 0), coordinates[1]] <- NA
    data[which(xydata$valid_coordinate == 0), coordinates[2]] <- NA
  }

  if (isTRUE(valid)) {
    data[, valid_name] <- xydata$valid_coordinate
  }

  return(data)
}
